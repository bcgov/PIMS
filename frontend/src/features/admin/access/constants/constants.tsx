import { ColumnWithProps } from 'components/Table';
import { AccessStatusDisplayMapper } from 'constants/accessStatus';
import * as React from 'react';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';
import { IAccessRequestModel } from '../interfaces';

export const columnDefinitions: ColumnWithProps<IAccessRequestModel>[] = [
  {
    Header: 'IDIR/BCeID',
    accessor: 'username',
    align: 'left',
    Cell: (props: CellProps<IAccessRequestModel>) => {
      return (
        <Link to={`/admin/user/${props.row.original.userId}`}>{props.row.original.username}</Link>
      );
    },
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Last name',
    accessor: 'lastName',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
  },
  {
    Header: 'Status',
    accessor: 'status',
    align: 'left',
    clickable: true,
    width: 100,
    Cell: (props: CellProps<IAccessRequestModel>) =>
      AccessStatusDisplayMapper[props.row.original.status],
  },
  {
    Header: 'Agency',
    accessor: 'agency',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Role',
    accessor: 'role',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: ' ',
    Cell: RowActions,
    width: 75,
  },
];
