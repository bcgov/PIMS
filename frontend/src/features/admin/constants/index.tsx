import * as React from 'react';
import { ColumnWithProps } from 'components/Table';
import { CellProps } from 'react-table';
import { Link } from 'react-router-dom';
import { IAgency, IAgencyRecord } from 'interfaces';

export const columnDefinitions: ColumnWithProps<IAgencyRecord>[] = [
  {
    Header: 'Agency name',
    accessor: 'name',
    align: 'left',
    sortable: true,
    clickable: true,
  },
  {
    Header: 'Short name',
    accessor: 'code',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Description',
    accessor: 'description',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Parent Agency',
    accessor: 'parent',
    align: 'left',
    clickable: true,
    Cell: (props: CellProps<IAgency>) => {
      return (
        <Link to={`/admin/agency/${props.row.original.parentId}`}>{props.row.original.parent}</Link>
      );
    },
    sortable: true,
  },
];
