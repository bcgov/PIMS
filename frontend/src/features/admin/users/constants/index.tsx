import { ColumnWithProps } from 'components/Table';
import * as React from 'react';
import { FaCheck, FaTimes } from 'react-icons/fa';
import { Link } from 'react-router-dom';
import { CellProps } from 'react-table';

import { RowActions } from '../components/RowActions';
import { IUserRecord } from '../interfaces/IUserRecord';

export const columnDefinitions: ColumnWithProps<IUserRecord>[] = [
  {
    Header: 'IDIR/BCeID',
    accessor: 'username',
    align: 'left',
    Cell: (props: CellProps<IUserRecord>) => {
      return <Link to={`/admin/user/${props.row.original.id}`}>{props.row.original.username}</Link>;
    },
    sortable: true,
  },
  {
    Header: 'First name',
    accessor: 'firstName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Last name',
    accessor: 'lastName',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Email',
    accessor: 'email',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 200,
  },
  {
    Header: 'Position',
    accessor: 'position',
    align: 'left',
    clickable: true,
    sortable: true,
  },
  {
    Header: 'Active',
    accessor: 'isDisabled',
    align: 'left',
    clickable: true,
    sortable: true,
    width: 100,
    Cell: (props: CellProps<IUserRecord>) =>
      props.row.original.isDisabled ? <FaTimes /> : <FaCheck />,
  },
  {
    Header: 'Agency',
    accessor: 'agency',
    align: 'left',
    sortable: true,
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Roles',
    accessor: 'roles',
    align: 'left',
    clickable: true,
    minWidth: 200,
  },
  {
    Header: 'Last Login',
    accessor: 'lastLogin',
    align: 'left',
    clickable: true,
    minWidth: 100,
  },
  {
    Header: 'Created On',
    accessor: 'createdOn',
    align: 'left',
    clickable: true,
    sortable: true,
    minWidth: 100,
  },
  {
    Header: ' ',
    Cell: RowActions,
    width: 75,
  },
];
