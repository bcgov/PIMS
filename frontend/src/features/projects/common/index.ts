export * from './interfaces';
export * from './components/FilterBar';
export * from './projectsActionCreator';
export { default as FilterBar } from './components/FilterBar';
export { default as StepStatusIcon } from './components/StepStatusIcon';
export { default as SresManual } from './components/SresManual';
export { default as StepErrorSummary } from './components/StepErrorSummary';
export { default as ProjectNotes, PrivateNotes, PublicNotes } from './components/ProjectNotes';
export { default as EditButton } from './components/EditButton';
export { default as SelectProjectPropertiesPage } from './components/SelectProjectPropertiesPage';
export { default as FormikTable } from './components/FormikTable';
export { default as ApprovalTransitionPage } from './ApprovalTransitionPage';

export { default as projectSlice } from './slices/projectSlice';
export { default as projectTasksSlice } from './slices/projectTasksSlice';
export { default as projectWorflowSlice } from './slices/projectWorkflowSlice';

export { default as ApprovalConfirmationForm } from './forms/ApprovalConfirmationForm';
export { default as UpdateInfoForm } from './forms/UpdateInfoForm';
export { default as DocumentationForm } from './forms/DocumentationForm';
export { default as ProjectDraftForm } from './forms/ProjectDraftForm';
export { default as AppraisalCheckListForm } from './forms/AppraisalCheckListForm';
export { default as FirstNationsCheckListForm } from './forms/FirstNationsCheckListForm';
export { default as SelectProjectPropertiesForm } from './forms/SelectProjectPropertiesForm';
export { default as TasksForm } from './forms/TasksForm';

export { default as useStepForm } from './hooks/useStepForm';
export { default as useProject } from './hooks/useProject';
export { default as useAgencyResponseTable } from './hooks/useAgencyResponseTable';

export * from './slices/projectSlice';
export * from './slices/projectTasksSlice';
export * from './slices/projectWorkflowSlice';

export * from './strings';

export { default as ProjectRouter } from './ProjectRouter';
export { default as ProjectSummaryView } from './ProjectSummaryView';

export { default as ProjectInformationTab } from './tabs/ProjectInformationTab';
export { default as DocumentationTab } from './tabs/DocumentationTab';
export { default as ErrorTabs } from './tabs/ErrorTabs';
export * from './tabs/tabValidation';
