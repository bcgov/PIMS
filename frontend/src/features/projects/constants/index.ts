export * from './AgencyResponses';
export * from './DisposalWorkflows';
export * from './DisposeWorkflowStatus';
export * from './ProjectWorkflowComponents';
export * from './ReviewWorkflowStatus';
export * from './SPPApprovalTabs';
