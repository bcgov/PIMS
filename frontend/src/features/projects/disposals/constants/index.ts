export * from './defaultFilter';
export * from './defaultProjectForm';
