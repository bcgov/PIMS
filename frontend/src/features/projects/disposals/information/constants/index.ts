export * from './AddPropertyColumns';
export * from './PropertyColumns';
