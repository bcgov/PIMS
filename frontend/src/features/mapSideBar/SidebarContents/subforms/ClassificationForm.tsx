import { FastSelect, SelectOptions, TextArea } from 'components/common/form';
import { Label } from 'components/common/Label';
import TooltipIcon from 'components/common/TooltipIcon';
import { Classifications } from 'constants/classifications';
import {
  CoreOperational,
  CoreStrategic,
  Demolished,
  Subdivided,
  SurplusActive,
  SurplusEncumbered,
  SurplusEncumberedOrActive,
  Disposed,
} from 'features/properties/components/forms/strings';
import { FormikValues, getIn, useFormikContext } from 'formik';
import React from 'react';
import { Col, Row } from 'react-bootstrap';
import styled from 'styled-components';
import variables from '_variables.module.scss';
import { useQueryParamSideBar, SidebarContextType } from '../../hooks/useQueryParamSideBar';

const Title = styled.h4`
  float: left;
`;

/** formated information box to display the classification definitions to the right of the select */
const InfoBox = styled.div`
  border: 1px solid ${variables.filterBackgroundColor};
  border-radius: 4px;
  text-align: left;
  padding: 8px 12px 10px;
  margin-bottom: 10px;
`;

/** formatted field description that will appear underneath the select input */
const FieldDescription = styled.p`
  margin-left: 166px;
  margin-top: 10px;
  font-size: 12px;
  text-align: left;
`;

const Encumberance = styled(Col)`
  .form-label {
    min-width: 140px;
  }
  .form-group {
    display: flex;
    textarea.form-control {
      width: 100%;
    }
  }
`;

/** used to conditionally render the information display box with the desired content */
const InfoBoxWithContent = (content: string) => {
  return (
    <InfoBox>
      <p style={{ fontSize: '11px' }}>{content}</p>
    </InfoBox>
  );
};

interface IClassificationFormProps {
  /** title to be displayed at the top of the form */
  title?: string;
  /** the classification set to be used with the component, could vary by form */
  classifications: SelectOptions;
  /** the field label for the select options (eg) Building Classification/Land Classification */
  fieldLabel?: string;
  /** pass down the appropriate field to edit */
  field: string;
  /** pass a tooltip for the classification field */
  toolTip?: string;
  /** reason for encumberance */
  encumbranceField: string;
  /** disable form controls */
  disabled?: boolean;
}

/**
 * component responsible for displaying classifcations of properties with conditional renders of information based on the classification selected.
 */
export const ClassificationForm: React.FC<IClassificationFormProps> = ({
  title,
  classifications,
  fieldLabel,
  field,
  toolTip,
  disabled,
  encumbranceField,
}) => {
  const formikProps = useFormikContext();

  const formikRef = React.useRef<FormikValues>();
  const { context } = useQueryParamSideBar(formikRef);
  let surplusActiveOrEncumbered =
    getIn(formikProps.values, field) === Classifications.SurplusEncumbered ||
    getIn(formikProps.values, field) === Classifications.SurplusActive;

  /** classId based on current formik values to determine which classsification information box to display */
  let classId = getIn(formikProps.values, field);

  /**
   * determine how to filter the classifications
   * @param id - The classification id.
   */
  const determineFilter = (id: Number) => {
    switch (context) {
      case SidebarContextType.ADD_BUILDING:
      case SidebarContextType.VIEW_BUILDING:
      case SidebarContextType.UPDATE_BUILDING:
        return id !== Classifications.Subdivided;
      case SidebarContextType.ADD_BARE_LAND:
      case SidebarContextType.VIEW_BARE_LAND:
      case SidebarContextType.UPDATE_BARE_LAND:
      case SidebarContextType.VIEW_DEVELOPED_LAND:
      case SidebarContextType.UPDATE_DEVELOPED_LAND:
        return id !== Classifications.Demolished;
      case SidebarContextType.ADD_SUBDIVISION_LAND:
      case SidebarContextType.VIEW_SUBDIVISION_LAND:
      case SidebarContextType.UPDATE_SUBDIVISION_LAND:
      case SidebarContextType.ADD_ASSOCIATED_LAND:
        return (
          id !== Classifications.Disposed &&
          id !== Classifications.Demolished &&
          id !== Classifications.Subdivided
        );
      default:
        return true;
    }
  };

  let filteredClassifications = classifications;
  /** users not allowed to select disposed, but display these values if one of these classifications has already been selected. */
  filteredClassifications = classifications.filter(
    c => determineFilter(+c.value) || +c.value === +classId,
  );

  const renderInfo = () => {
    switch (classId) {
      case Classifications.CoreOperational:
        return InfoBoxWithContent(CoreOperational);
      case Classifications.CoreStrategic:
        return InfoBoxWithContent(CoreStrategic);
      case Classifications.SurplusEncumbered:
        return InfoBoxWithContent(SurplusEncumbered);
      case Classifications.SurplusActive:
        return InfoBoxWithContent(SurplusActive);
      case Classifications.Demolished:
        return InfoBoxWithContent(Demolished);
      case Classifications.Subdivided:
        return InfoBoxWithContent(Subdivided);
      case Classifications.Disposed:
        return InfoBoxWithContent(Disposed);
      default:
        return InfoBoxWithContent(
          'Select a classification from the dropdown list to show the definition here. For further information, see the Inventory Policy.',
        );
    }
  };

  return (
    <>
      <Row>
        <Title>{title}</Title>
      </Row>
      <Row>
        <Col md={6}>
          <Row style={{ display: 'flex' }}>
            <Label style={{ marginBottom: '0', textAlign: 'right' }}>{fieldLabel}</Label>
            <FastSelect
              formikProps={formikProps}
              type="number"
              style={{ marginTop: '5px', display: 'flex' }}
              placeholder="Must Select One"
              field={field}
              options={filteredClassifications}
              disabled={disabled}
              required
              displayErrorTooltips
            />
            {toolTip && (
              <div style={{ marginTop: '8px' }}>
                <TooltipIcon toolTip={toolTip} toolTipId="classificationToolTip" />
              </div>
            )}
          </Row>
        </Col>
        <Col md={6}>{renderInfo()}</Col>
      </Row>
      {getIn(formikProps.values, field) === Classifications.SurplusEncumbered && (
        <Row>
          <Encumberance>
            <TextArea field={encumbranceField} label="Reason for Encumbrance"></TextArea>
          </Encumberance>
        </Row>
      )}
      {surplusActiveOrEncumbered && (
        <FieldDescription>{SurplusEncumberedOrActive}</FieldDescription>
      )}
    </>
  );
};
