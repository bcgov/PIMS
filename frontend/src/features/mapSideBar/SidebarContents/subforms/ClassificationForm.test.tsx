import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, fireEvent, render, wait } from '@testing-library/react';
import { Formik } from 'formik';
import { noop } from 'lodash';
import axios from 'axios';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import { useKeycloak } from '@react-keycloak/web';
import { ClassificationForm } from './ClassificationForm';
import { Classifications } from 'constants/classifications';
import { SelectOption, SelectOptions } from 'components/common/form';
import * as API from 'constants/API';
import * as reducerTypes from 'constants/reducerTypes';
import { ILookupCode } from 'actions/lookupActions';

jest.mock('axios');
jest.mock('@react-keycloak/web');

const mockedAxios = axios as jest.Mocked<typeof axios>;
const mockClassifications = [
  { value: Classifications.CoreOperational, label: 'Core Operational' } as SelectOption,
  { value: Classifications.CoreStrategic, label: 'Core Strategic' } as SelectOption,
  { value: Classifications.SurplusEncumbered, label: 'Surplus Encumbered' } as SelectOption,
] as SelectOptions;

const mockKeycloak = (claims: string[]) => {
  (useKeycloak as jest.Mock).mockReturnValue({
    keycloak: {
      subject: 'test',
      userInfo: {
        roles: claims,
        agencies: ['1'],
      },
    },
  });
};
const mockStore = configureMockStore([thunk]);
let history = createMemoryHistory();

const lCodes = {
  lookupCodes: [
    { name: 'agencyVal', id: '1', isDisabled: false, type: API.AGENCY_CODE_SET_NAME },
    { name: 'disabledAgency', id: '2', isDisabled: true, type: API.AGENCY_CODE_SET_NAME },
    { name: 'roleVal', id: '1', isDisabled: false, type: API.ROLE_CODE_SET_NAME },
    { name: 'disabledRole', id: '2', isDisabled: true, type: API.ROLE_CODE_SET_NAME },
    {
      name: 'Core Operational',
      id: '0',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
    {
      name: 'Core Strategic',
      id: '1',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
    {
      name: 'Surplus Active',
      id: '2',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
    {
      name: 'Surplus Encumbered',
      id: '3',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
    {
      name: 'Disposed',
      id: '4',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
    {
      name: 'Demolished',
      id: '5',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
    {
      name: 'Subdivided',
      id: '6',
      isDisabled: false,
      type: API.PROPERTY_CLASSIFICATION_CODE_SET_NAME,
    },
  ] as ILookupCode[],
};

const getStore = () =>
  mockStore({
    [reducerTypes.PROPERTY_NAMES]: ['test'],
    [reducerTypes.LOOKUP_CODE]: lCodes,
  });

it('renders correctly', () => {
  const tree = renderer.create(form).toJSON();
  expect(tree).toMatchSnapshot();
});

const form = (
  <Provider store={getStore()}>
    <Router history={history}>
      <Formik onSubmit={noop} initialValues={{ classificationId: '' }}>
        <ClassificationForm
          classifications={mockClassifications}
          field="classificationId"
          encumbranceField="test"
        />
      </Formik>
    </Router>
  </Provider>
);

describe('renders definitions correctly', () => {
  mockedAxios.get.mockImplementationOnce(() => Promise.resolve({}));

  beforeEach(() => {
    mockKeycloak([]);
  });
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });
  it('renders default defintion', () => {
    const { getByText } = render(form);
    expect(getByText(/select a classification/i)).toBeInTheDocument();
  });
  it('renders core operational definition', async () => {
    const { getByText, container } = render(form);
    const classificationId = container.querySelector('select[name="classificationId"]');
    await wait(() => {
      fireEvent.change(classificationId!, {
        target: {
          value: Classifications.CoreOperational,
        },
      });
    });
    expect(getByText(/core Operational – assets that are functionally/i)).toBeInTheDocument();
  });
  it('renders core strategic definition', async () => {
    const { getByText, container } = render(form);
    const classificationId = container.querySelector('select[name="classificationId"]');
    await wait(() => {
      fireEvent.change(classificationId!, {
        target: {
          value: Classifications.CoreStrategic,
        },
      });
    });
    expect(getByText(/core Strategic – assets that are uniquely/i)).toBeInTheDocument();
  });
  it('renders surplus encumbered definition', async () => {
    const { getByText, container } = render(form);
    const classificationId = container.querySelector('select[name="classificationId"]');
    await wait(() => {
      fireEvent.change(classificationId!, {
        target: {
          value: Classifications.SurplusEncumbered,
        },
      });
    });
    expect(getByText(/surplus Encumbered – assets that are surplus/i)).toBeInTheDocument();
  });
});
