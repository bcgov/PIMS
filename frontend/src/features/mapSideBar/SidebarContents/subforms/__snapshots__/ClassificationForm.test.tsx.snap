// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  .c0 {
  float: left;
}

<div
    className="row"
  >
    <h4
      className="c0"
    />
  </div>,
  .c0 {
  border: 1px solid #f2f2f2;
  border-radius: 4px;
  text-align: left;
  padding: 8px 12px 10px;
}

<div
    className="row"
  >
    <div
      className="col-md-6"
    >
      <div
        className="row"
        style={
          Object {
            "display": "flex",
          }
        }
      >
        <p
          className="label"
          style={
            Object {
              "marginBottom": "0",
              "textAlign": "right",
            }
          }
        />
        <div
          className="required form-group"
        >
          <select
            className="form-control"
            id="input-classificationId-fast"
            name="classificationId"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            onMouseOut={[Function]}
            onMouseOver={[Function]}
            required={true}
            style={
              Object {
                "display": "flex",
                "marginTop": "5px",
              }
            }
            type="number"
            value=""
          >
            <option
              value=""
            >
              Must Select One
            </option>
            <option
              value={0}
            >
              Core Operational
            </option>
            <option
              value={1}
            >
              Core Strategic
            </option>
            <option
              value={3}
            >
              Surplus Encumbered
            </option>
          </select>
        </div>
      </div>
    </div>
    <div
      className="col-md-6"
    >
      <div
        className="c0"
      >
        <p
          style={
            Object {
              "fontSize": "11px",
            }
          }
        >
          Select a classification from the dropdown list to show the definition here. For further information, see the Inventory Policy.
        </p>
      </div>
    </div>
  </div>,
]
`;
